mopidy/audio/scan.py

```diff
@@ -68,117 +80,144 @@ def scan(self, uri, timeout=None):
+    decodebin = Gst.ElementFactory.make('decodebin')
 
-    pipeline = gst.element_factory_make('pipeline')
+    pipeline = Gst.ElementFactory.make('pipeline')
     for e in (src, typefind, decodebin):
         pipeline.add(e)
-    gst.element_link_many(src, typefind, decodebin)
+    src.link(typefind)
+    typefind.link(decodebin)
 
     if proxy_config:
         utils.setup_proxy(src, proxy_config)
 
-    typefind.connect('have-type', _have_type, decodebin)
-    decodebin.connect('pad-added', _pad_added, pipeline)
+    signals = utils.Signals()
+    signals.connect(typefind, 'have-type', _have_type, decodebin)
+    signals.connect(decodebin, 'pad-added', _pad_added, pipeline)
+    signals.connect(decodebin, 'autoplug-select', _autoplug_select)
 
-    return pipeline
+    return pipeline, signals
 
 
 def _have_type(element, probability, caps, decodebin):
     decodebin.set_property('sink-caps', caps)
-    struct = gst.Structure('have-type')
-    struct['caps'] = caps.get_structure(0)
-    element.get_bus().post(gst.message_new_application(element, struct))
+    struct = Gst.Structure.new_empty('have-type')
+    struct.set_value('caps', caps.get_structure(0))
+    element.get_bus().post(Gst.Message.new_application(element, struct))
 
 
 def _pad_added(element, pad, pipeline):
-    sink = gst.element_factory_make('fakesink')
+    sink = Gst.ElementFactory.make('fakesink')
     sink.set_property('sync', False)
 
     pipeline.add(sink)
     sink.sync_state_with_parent()
-    pad.link(sink.get_pad('sink'))
+    pad.link(sink.get_static_pad('sink'))
 
-    if pad.get_caps().is_subset(_RAW_AUDIO):
-        struct = gst.Structure('have-audio')
-        element.get_bus().post(gst.message_new_application(element, struct))
+    if pad.query_caps().is_subset(Gst.Caps.from_string('audio/x-raw')):
+        # Probably won't happen due to autoplug-select fix, but lets play it
+        # safe until we've tested more.
+        struct = Gst.Structure.new_empty('have-audio')
+        element.get_bus().post(Gst.Message.new_application(element, struct))
+
+
+def _autoplug_select(element, pad, caps, factory):
+    if factory.list_is_type(_DECODER | _AUDIO):
+        struct = Gst.Structure.new_empty('have-audio')
+        element.get_bus().post(Gst.Message.new_application(element, struct))
+    if not factory.list_is_type(_DEMUXER | _DEPAYLOADER | _PARSER):
+        return _SELECT_EXPOSE
+    return _SELECT_TRY
 
 
 def _start_pipeline(pipeline):
-    if pipeline.set_state(gst.STATE_PAUSED) == gst.STATE_CHANGE_NO_PREROLL:
-        pipeline.set_state(gst.STATE_PLAYING)
+    result = pipeline.set_state(Gst.State.PAUSED)
+    if result == Gst.StateChangeReturn.NO_PREROLL:
+        pipeline.set_state(Gst.State.PLAYING)
 
 
-def _query_duration(pipeline):
-    try:
-        duration = pipeline.query_duration(gst.FORMAT_TIME, None)[0]
-    except gst.QueryError:
-        return None
+def _query_duration(pipeline, timeout=100):
```

Comment:
I probably should have commented this code better as it probably won't be clear to anyone why we go playing to get the duration in this way.